From 657f7b37d2f46fcbd201dc5e258dc0b07d493ab1 Mon Sep 17 00:00:00 2001
From: Ben Gao <bengao168@msn.com>
Date: Tue, 22 Aug 2023 16:11:33 +0800
Subject: [PATCH] bugfix: webCrawl can't handle website well recursively When
 using Cheerio Web Scraper and selecting Web Crawl, it is supposed to crawling
 web pages recursively. There were couple bugs for successfully doing that: 1,
 extracting urls from webpage with things like default:blank and mailto etc.
 2, extracting rules are not well aligned with specification of html <a>
 element 3, it always returns 10 pages of the website Furthermore, this commit
 also adds a few informative logs when debug enabled

Signed-off-by: Ben Gao <bengao168@msn.com>
---
 .../nodes/documentloaders/Cheerio/Cheerio.ts  |  1 +
 packages/components/src/utils.ts              | 62 ++++++++++++-------
 2 files changed, 42 insertions(+), 21 deletions(-)

diff --git a/packages/components/nodes/documentloaders/Cheerio/Cheerio.ts b/packages/components/nodes/documentloaders/Cheerio/Cheerio.ts
index 1c21c1ea867..e5b9523ed43 100644
--- a/packages/components/nodes/documentloaders/Cheerio/Cheerio.ts
+++ b/packages/components/nodes/documentloaders/Cheerio/Cheerio.ts
@@ -113,6 +113,7 @@ class Cheerio_DocumentLoaders implements INode {
             if (process.env.DEBUG === 'true') console.info(`pages: ${JSON.stringify(pages)}, length: ${pages.length}`)
             if (!pages || pages.length === 0) throw new Error('No relative links found')
             for (const page of pages) {
+		if (process.env.DEBUG === 'true') console.info(`loading page: ${page}`)
                 docs.push(...(await cheerioLoader(page)))
             }
             if (process.env.DEBUG === 'true') console.info(`Finish ${relativeLinksMethod}`)
diff --git a/packages/components/src/utils.ts b/packages/components/src/utils.ts
index 8d06a6501bd..74a1ab52667 100644
--- a/packages/components/src/utils.ts
+++ b/packages/components/src/utils.ts
@@ -214,24 +214,44 @@ function getURLsFromHTML(htmlBody: string, baseURL: string): string[] {
     const dom = new JSDOM(htmlBody)
     const linkElements = dom.window.document.querySelectorAll('a')
     const urls: string[] = []
+	const hostWithPort = baseURL.split('/')[2]
+	const protocol = baseURL.split('/')[0]
+	if (process.env.DEBUG === 'true') console.error(`enter getURLsFromHTML with baseURL: ${baseURL}`)
     for (const linkElement of linkElements) {
-        if (linkElement.href.slice(0, 1) === '/') {
-            try {
-                const urlObj = new URL(baseURL + linkElement.href)
-                urls.push(urlObj.href) //relative
-            } catch (err) {
-                if (process.env.DEBUG === 'true') console.error(`error with relative url: ${err.message}`)
-                continue
-            }
-        } else {
-            try {
-                const urlObj = new URL(linkElement.href)
-                urls.push(urlObj.href) //absolute
-            } catch (err) {
-                if (process.env.DEBUG === 'true') console.error(`error with absolute url: ${err.message}`)
-                continue
-            }
-        }
+		if(!linkElement.href || 
+			linkElement.href.startsWith("about:blank") || 
+			linkElement.href.startsWith("mailto:") ||
+			linkElement.href.indexOf('#') != -1) continue
+		if(linkElement.href.startsWith('http://') || 
+			linkElement.href.startsWith('https://') || 
+			linkElement.href.startsWith('//')) {
+			try {
+				if(linkElement.href.startsWith('http')) {
+					const urlObj = new URL(linkElement.href)
+					urls.push(urlObj.href)
+				}
+				else {
+					const urlObj = new URL(protocol + linkElement.href)
+					urls.push(urlObj.href)
+				}
+			}
+			catch(err) {	
+				if (process.env.DEBUG === 'true') console.error(`error with absolute url: ${err.message} with href ${linkElement.href}`)
+				continue
+			}
+		}
+		else {
+			try {
+				const urlObj = new URL((linkElement.href.startsWith('/') ? 
+										protocol + "//" + hostWithPort + linkElement.href : 
+				                        baseURL + '/' + linkElement.href))
+				urls.push(urlObj.href)
+			}
+			catch(err) {
+				if (process.env.DEBUG === 'true') console.error(`error with relative url: ${err.message} with href ${linkElement.href}`)
+				continue
+			}
+		}
     }
     return urls
 }
@@ -262,7 +282,7 @@ function normalizeURL(urlString: string): string {
 async function crawl(baseURL: string, currentURL: string, pages: string[], limit: number): Promise<string[]> {
     const baseURLObj = new URL(baseURL)
     const currentURLObj = new URL(currentURL)
-
+    if (process.env.DEBUG === 'true') console.info(`enter crawl: ${baseURL} ${currentURL}`)
     if (limit !== 0 && pages.length === limit) return pages
 
     if (baseURLObj.hostname !== currentURLObj.hostname) return pages
@@ -274,7 +294,7 @@ async function crawl(baseURL: string, currentURL: string, pages: string[], limit
 
     pages.push(normalizeCurrentURL)
 
-    if (process.env.DEBUG === 'true') console.info(`actively crawling ${currentURL}`)
+    if (process.env.DEBUG === 'true') console.info(`actively crawling ${currentURL} pages: ${pages.length}`)
     try {
         const resp = await fetch(currentURL)
 
@@ -290,9 +310,9 @@ async function crawl(baseURL: string, currentURL: string, pages: string[], limit
         }
 
         const htmlBody = await resp.text()
-        const nextURLs = getURLsFromHTML(htmlBody, baseURL)
+        const nextURLs = getURLsFromHTML(htmlBody, normalizeCurrentURL)
         for (const nextURL of nextURLs) {
-            pages = await crawl(baseURL, nextURL, pages, limit)
+            pages = await crawl(normalizeCurrentURL, nextURL, pages, limit)
         }
     } catch (err) {
         if (process.env.DEBUG === 'true') console.error(`error in fetch url: ${err.message}, on page: ${currentURL}`)
